services/kafka/hash.go

```diff
@@ -0,0 +1,116 @@
+package kafka
+
+import (
+	"errors"
+	"hash/crc32"
+	"math/rand"
+	"time"
+
+	"github.com/Shopify/sarama"
+)
+
+var ErrNonPositivePartitions = errors.New("number of partitions must be positive")
+
+type murmur2 struct{}
+
+// Partition here is based on the murmur2 balance function in github.com/segmentio/kafka-go to not break compatibility
+// with earlier versions.
+func (b murmur2) Partition(msg *sarama.ProducerMessage, numPartitions int32) (int32, error) {
+	if numPartitions == 0 {
+		return 0, ErrNonPositivePartitions
+	}
+
+	// NOTE: the murmur2 balancers in java and librdkafka treat a nil key as
+	//       non-existent while treating an empty slice as a defined value.
```

Comment:
Is this consistent with the previous library behaviour that kapa was using? nil vs empty slices in go don't often have meaning, so I'm suprised the sarama parsing code would make a distinction, regardless of what the java representation is.

Reply:
The library we were using before made that distinction, because they were trying to mimic the default java client behavior.